### PR TITLE
Update alert notify template to support when alertmanager send multiple alert and add workload info for pod and event alert

### PR DESCRIPTION
**Pull request by aiwantaozi**
Update alert notify template to support when alertmanager send multi alert in one request

Problem:
1. Alertmanager may send alert array, current notify template just use the first alert in array to contribute the notify info, not support multi alert well
2. it would help user to know the alerting pod belong to which workload if we could add workload info
3. alert name is miss for node alert

Solution:
1. Update notify template to support multi alert
2. add workload info 
3. add alert name in node alert

Issue:
https://github.com/rancher/rancher/issues/17472
https://github.com/rancher/rancher/issues/16990
https://github.com/rancher/rancher/issues/17535

### CODE REVIEW

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 1]
```diff
@@ -512,3 +528,29 @@ func includeProjectMetrics(projectAlerts []*v3.ProjectAlertRule) bool {
 	}
 	return false
 }
+
+func getClusterAlertGroupBy(spec v3.ClusterAlertRuleSpec) []model.LabelName {
+	if spec.EventRule != nil {
+		return eventGroupBy
```
Comment (orangedeng) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: If you don't need to reuse the variable `eventGroupBy `, just return a new array. Same as all returns in this function.

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 2]
```diff
@@ -512,3 +528,29 @@ func includeProjectMetrics(projectAlerts []*v3.ProjectAlertRule) bool {
...
+	} else if spec.MetricRule != nil {
+		return metricGroupBy
+	}
+
+	return nil
+}
+
+func getProjectAlertGroupBy(spec v3.ProjectAlertRuleSpec) []model.LabelName {
+	if spec.PodRule != nil {
+		return podGroupBy
```
Comment (orangedeng) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: If you don't need to reuse the variable `podGroupBy `, just return a new array. Same as all returns in this function.

File: pkg/controllers/user/alert/configsyncer/configsyncer.go [hunk 3]
```diff
@@ -512,3 +518,29 @@ func includeProjectMetrics(projectAlerts []*v3.ProjectAlertRule) bool {
...
+	}
+
+	return nil
+}
+
+func getProjectAlertGroupBy(spec v3.ProjectAlertRuleSpec) []model.LabelName {
+	if spec.PodRule != nil {
+		return []model.LabelName{"rule_id", "namespace", "pod_name"}
+	} else if spec.WorkloadRule != nil {
+		return []model.LabelName{"rule_id", "workload_namespace", "workload_name"}
```
Comment (orangedeng) [inline pkg/controllers/user/alert/configsyncer/configsyncer.go]: I think you will need workload kind for this.

File: pkg/controllers/user/alert/deployer/notification_template.go [hunk 4]
```diff
@@ -1,229 +1,254 @@
...
+{{ range .Alerts.Firing }}
+{{ template "__text_single" . }}
+{{ end -}}
+{{ end -}}
+
+
+
+{{- define "__text_single" -}}
+{{- if eq .Labels.alert_type "event" }}
+Alert Name: {{ .Labels.alert_name}}
```
Comment (orangedeng) [inline pkg/controllers/user/alert/deployer/notification_template.go]: I see so many following blocks in the template. Should we move them to the top of the `__text_single` template. 
```
Alert Name: ...
Severity: ...
Cluster Name: ...
```

File: pkg/controllers/user/alert/watcher/event.go [hunk 5]
```diff
@@ -91,3 +113,33 @@ func (l *EventWatcher) Sync(key string, obj *corev1.Event) (runtime.Object, erro
 
 	return nil, nil
 }
+
+type workloadFetcher struct {
+	unversionedClient rest.Interface
+}
+
+func (w *workloadFetcher) getOwnerRefferences(name, namespace, kind string) (rootName string, err error) {
```
Comment (orangedeng) [inline pkg/controllers/user/alert/watcher/event.go]: Is it a typo?

File: pkg/controllers/user/alert/watcher/event.go [hunk 6]
```diff
@@ -91,3 +113,33 @@ func (l *EventWatcher) Sync(key string, obj *corev1.Event) (runtime.Object, erro
...
+	obj := &unstructured.Unstructured{}
+	objKind := nameutil.GuessPluralName(kind)
+	objPrefix := resourcelink.ExportResourcePrefixMappings[strings.ToLower(objKind)]
+	if err = w.unversionedClient.Get().Prefix(objPrefix).Namespace(namespace).Resource(objKind).Name(name).Do().Into(obj); err != nil {
+		return "", errors.Wrapf(err, "get %s %s:%s failed", objKind, namespace, name)
+	}
+
+	allRef := obj.GetOwnerReferences()
+	if len(allRef) == 0 {
```
Comment (orangedeng) [inline pkg/controllers/user/alert/watcher/event.go]: There is a chance for workload which also has the owner reference information. Like the prometheus statefulSet from prometheus operator.

**Comment by cjellick** [comment]
LGTM

**Pull request merged**